qutip/models/dicke.py

```diff
@@ -0,0 +1,1325 @@
+    if (not float(N).is_integer()):
+        raise ValueError("Number of TLS should be an integer")
+
+    if (N < 1):
+        raise ValueError("Number of TLS should be non-negative")
+
+    nds = (N / 2 + 1)**2 - (N % 2) / 4
+    return int(nds)
+
+
+def num_dicke_ladders(N):
+    """
+    Calculates the total number of Dicke ladders in the Dicke space for a
+    collection of N two-level systems. It counts how many different "j" exist.
+    Or the number of blocks in the block diagonal matrix.
+
+    Parameters
+    -------
+    N: int
+        The number of two level systems.
+    Returns
+    -------
+    Nj: int
+        The number of Dicke ladders
+    """
+    Nj = (N + 1) * 0.5 + (1 - np.mod(N, 2)) * 0.5
+    return int(Nj)
+
+
+def num_tls(nds):
+    """
+    The number of two level systems (TLS), given the number of Dicke states.
+    Inverse function of _num_dicke_states(N)
+
+    Parameters
+    ----------
+    nds: int
+         The number of Dicke states
+    Returns
+    -------
+    N: int
+        The number of two level systems
+    """
+    if np.sqrt(nds).is_integer():
+        # N is even
+        N = 2 * (np.sqrt(nds) - 1)
+    else:
+        # N is odd
+        N = 2 * (np.sqrt(nds + 1 / 4) - 1)
+
+    return int(N)
+
+
+def isdiagonal(matrix):
+    """
+    Check if a matrix is diagonal either if it is a Qobj or a ndarray
+    """
+    if isinstance(matrix, Qobj):
+        matrix = matrix.full()
+
+    isdiag = np.all(matrix == np.diag(np.diagonal(matrix)))
+
+    return isdiag
+
+
+def dicke_states(N):
+    """
+    The number of dicke states with a modulo term taking care of ensembles
+    with odd number of systems. Same function, named "num_dicke_states",
+    is present in the file with cythonized code, dicke.pyx. 
+
+    Parameters
+    -------
+    N: int
+        The number of two level systems
+    Returns
+    -------
+    nds: int
+        The number of Dicke states
+    """
+    if (not float(N).is_integer()):
+        raise ValueError("Number of TLS should be an integer")
+
+    if (N < 1):
+        raise ValueError("Number of TLS should be non-negative")
+
+    nds = (N / 2 + 1)**2 - (N % 2) / 4
+
+    return int(nds)
+
+def dicke_ladders(N):
+    """
+    Calculates the total number of Dicke ladders in the Dicke space for a
+    collection of N two-level systems. It counts how many different "j" exist.
+    Or the number of blocks in the block diagonal matrix. Same function, 
+    named "num_dicke_ladders", is present in the file with cythonized code, dicke.pyx.
+
+    Parameters
+    -------
+    N: int
+        The number of two level systems.
+    Returns
+    -------
+    Nj: int
+        The number of Dicke ladders
+    """
+    Nj = (N + 1) * 0.5 + (1 - np.mod(N, 2)) * 0.5
+
+    return int(Nj)
+
+class Piqs(object):
+    """
+    The Dicke States class.
+
+    Parameters
+    ----------
+    N : int
+        The number of two level systems
+        default: 2
+
+    hamiltonian : Qobj matrix
```

Comment:
If you give the qobj type as 
```
    hamiltonian:  :class:`qutip.Qobj`
```
then it will link in the documentation.
There are lots of other examples. I won't tag them all